File: backend/app/core/kinematics.py
"""3D Kinematics and RULA Ergonomic Evaluation Engine."""
from typing import Tuple, Dict, Any
import math


def euclidean_distance_3d(
    p1: Tuple[float, float, float], p2: Tuple[float, float, float]
) -> float:
    """Calculates Euclidean distance between two 3D points."""
    return math.sqrt(
        (p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2 + (p1[2] - p2[2]) ** 2
    )


def calculate_cva_3d(
    tragus: Tuple[float, float, float], c7_mid_shoulder: Tuple[float, float, float]
) -> float:
    """Calculates the Craniovertebral Angle (CVA) in 3D world space.

    Formed by the horizontal plane through C7/mid-shoulder and the tragus-to-C7 line.
    Normal upright CVA is >= 50 degrees; forward head posture drops < 45-50 degrees.
    """
    dx = tragus[0] - c7_mid_shoulder[0]
    dy = c7_mid_shoulder[1] - tragus[1]  # In standard coordinates, head Y is higher (smaller value)
    dz = tragus[2] - c7_mid_shoulder[2]

    # Horizontal plane projection distance
    d_horizontal = math.sqrt(dx * dx + dz * dz)
    if d_horizontal < 1e-6:
        return 90.0

    angle_rad = math.atan2(abs(dy), d_horizontal)
    cva_deg = math.degrees(angle_rad)
    return round(cva_deg, 1)


def calculate_shoulder_tilt_3d(
    left_shoulder: Tuple[float, float, float], right_shoulder: Tuple[float, float, float]
) -> float:
    """Calculates horizontal tilt angle between left and right shoulders."""
    dx = right_shoulder[0] - left_shoulder[0]
    dy = right_shoulder[1] - left_shoulder[1]
    if abs(dx) < 1e-6:
        return 90.0
    tilt_rad = math.atan(dy / dx)
    return round(abs(math.degrees(tilt_rad)), 1)


def calculate_trunk_angle_3d(
    mid_hip: Tuple[float, float, float], mid_shoulder: Tuple[float, float, float]
) -> float:
    """Calculates spine trunk angle deviation relative to the vertical axis (0, 1, 0)."""
    dx = mid_shoulder[0] - mid_hip[0]
    dy = mid_hip[1] - mid_shoulder[1]  # Height
    dz = mid_shoulder[2] - mid_hip[2]

    norm = math.sqrt(dx * dx + dy * dy + dz * dz)
    if norm < 1e-6:
        return 0.0

    # Angle with vertical axis
    cos_theta = max(-1.0, min(1.0, abs(dy) / norm))
    trunk_angle_rad = math.acos(cos_theta)
    return round(math.degrees(trunk_angle_rad), 1)


def compute_rula_score(
    cva_deg: float, trunk_angle_deg: float, shoulder_tilt_deg: float
) -> Dict[str, Any]:
    """Maps Craniovertebral Angle (CVA), trunk angle, and shoulder tilt to standard RULA scores (1–7).

    Returns:
        {
            "rula_score": int (1..7),
            "risk_level": str,
            "action_required": str,
            "color": str
        }
    """
    # 1. Neck score from CVA
    if cva_deg >= 52.0:
        neck_score = 1
    elif cva_deg >= 45.0:
        neck_score = 2
    elif cva_deg >= 38.0:
        neck_score = 3
    else:
        neck_score = 4

    # 2. Trunk flexion score
    if trunk_angle_deg <= 8.0:
        trunk_score = 1
    elif trunk_angle_deg <= 16.0:
        trunk_score = 2
    elif trunk_angle_deg <= 28.0:
        trunk_score = 3
    else:
        trunk_score = 4

    # 3. Shoulder / Upper body alignment score
    if shoulder_tilt_deg <= 4.0:
        shoulder_score = 1
    elif shoulder_tilt_deg <= 10.0:
        shoulder_score = 2
    else:
        shoulder_score = 3

    # Composite RULA mapping
    combined = max(neck_score, trunk_score)
    if shoulder_score > 1:
        combined += 1
    if neck_score >= 3 and trunk_score >= 3:
        combined += 1

    rula = max(1, min(7, combined))

    if rula <= 2:
        risk_level = "Acceptable"
        action = "Posture is acceptable if not maintained for prolonged periods."
        color = "#10b981"  # Emerald
    elif rula <= 4:
        risk_level = "Further Investigation"
        action = "Further investigation needed; minor adjustments recommended."
        color = "#f59e0b"  # Amber
    elif rula <= 6:
        risk_level = "Change Soon"
        action = "Investigation and changes required soon to avoid strain."
        color = "#f97316"  # Orange
    else:
        risk_level = "Immediate Action"
        action = "Immediate postural intervention and ergonomic changes required."
        color = "#ef4444"  # Rose/Red

    return {
        "rula_score": rula,
        "risk_level": risk_level,
        "action_required": action,
        "color": color,
        "sub_scores": {
            "neck": neck_score,
            "trunk": trunk_score,
            "shoulder": shoulder_score,
        },
    }


def calculate_ipd_proximity(
    current_ipd: float, baseline_ipd: float
) -> Dict[str, Any]:
    """Calculates screen proximity ratio using Inter-Pupillary Distance (IPD).

    Alert triggers if IPD increases by > 25% (indicates leaning closer than ~45 cm).
    """
    if baseline_ipd <= 1e-6:
        ratio = 1.0
    else:
        ratio = round(current_ipd / baseline_ipd, 2)

    too_close = ratio > 1.25  # Leaning closer than 45cm

    # Estimate viewing distance in centimeters (assuming 60cm baseline)
    estimated_distance_cm = round(60.0 / max(0.5, ratio), 1)

    return {
        "ipd_ratio": ratio,
        "too_close": too_close,
        "estimated_distance_cm": estimated_distance_cm,
        "warning": "Screen proximity too close (< 45cm)! Sit back." if too_close else None,
    }

File: frontend/src/core/kinematics.ts
/**
 * 3D Kinematics and Biomechanical RULA Ergonomic Evaluator
 * Computes Craniovertebral Angle (CVA), Shoulder Tilt, Spine Trunk Angle,
 * and standard RULA (1-7) Risk Assessment Score.
 */

import type { Point3D, RulaEvaluation, RulaLevel, PostureStatus } from '../types/ergosense';

/**
 * 3D Euclidean distance between two points.
 */
export function distance3D(p1: Point3D, p2: Point3D): number {
  const dx = p1.x - p2.x;
  const dy = p1.y - p2.y;
  const dz = (p1.z || 0) - (p2.z || 0);
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

/**
 * Midpoint between two 3D coordinates.
 */
export function midpoint3D(p1: Point3D, p2: Point3D): Point3D {
  return {
    x: (p1.x + p2.x) / 2,
    y: (p1.y + p2.y) / 2,
    z: ((p1.z || 0) + (p2.z || 0)) / 2,
    visibility: Math.min(p1.visibility ?? 1, p2.visibility ?? 1),
  };
}

/**
 * Calculates Craniovertebral Angle (CVA) in degrees.
 * Normal upright posture: 50° - 55°
 * Forward head slouch: < 48° (Severe: < 40°)
 *
 * @param tragus Ear tragus coordinate (left or right ear)
 * @param c7 C7 spinous process coordinate (approximated from shoulder midpoint)
 */
export function computeCVA(tragus: Point3D, c7: Point3D): number {
  // In screen/webcam coordinates, y increases downward.
  // Ear is above C7, so delta_y = c7.y - tragus.y is positive.
  const deltaY = Math.max(0.001, c7.y - tragus.y);
  const deltaX = tragus.x - c7.x;
  const deltaZ = (tragus.z || 0) - (c7.z || 0);

  // Horizontal plane distance
  const horizontalDist = Math.sqrt(deltaX * deltaX + deltaZ * deltaZ);

  // Angle with horizontal plane
  const angleRad = Math.atan2(deltaY, Math.max(0.0001, horizontalDist));
  const angleDeg = (angleRad * 180) / Math.PI;

  return Math.min(90, Math.max(10, Math.round(angleDeg * 10) / 10));
}

/**
 * Calculates horizontal shoulder tilt asymmetry in degrees.
 * Upright level: < 4°
 * Moderate tilt: 5° - 10°
 * High tilt: > 10°
 */
export function computeShoulderTilt(leftShoulder: Point3D, rightShoulder: Point3D): number {
  const deltaY = Math.abs(leftShoulder.y - rightShoulder.y);
  const deltaX = leftShoulder.x - rightShoulder.x;
  const deltaZ = (leftShoulder.z || 0) - (rightShoulder.z || 0);

  const horizontalSpan = Math.sqrt(deltaX * deltaX + deltaZ * deltaZ);
  if (horizontalSpan < 0.001) return 0.0;

  const angleRad = Math.atan2(deltaY, horizontalSpan);
  const angleDeg = (angleRad * 180) / Math.PI;

  return Math.round(angleDeg * 10) / 10;
}

/**
 * Calculates spine trunk inclination angle with true vertical (0, -1, 0).
 * Upright neutral: < 10°
 * Leaning/hunched: > 15° - 20°
 */
export function computeTrunkAngle(midShoulder: Point3D, midHip: Point3D): number {
  // Vector from mid hip to mid shoulder (should point upward: negative y)
  const vx = midShoulder.x - midHip.x;
  const vy = midShoulder.y - midHip.y; // vy < 0 when shoulder is above hip
  const vz = (midShoulder.z || 0) - (midHip.z || 0);

  const mag = Math.sqrt(vx * vx + vy * vy + vz * vz);
  if (mag < 0.001) return 0.0;

  // Dot product with vertical unit vector (0, -1, 0) is -vy
  const cosTheta = Math.max(-1.0, Math.min(1.0, -vy / mag));
  const thetaDeg = (Math.acos(cosTheta) * 180) / Math.PI;

  return Math.round(thetaDeg * 10) / 10;
}

/**
 * Evaluates comprehensive RULA Ergonomic Risk Score (1 to 7) based on biomechanical angles.
 *
 * Scoring Scale:
 * 1-2: Acceptable (Green #10b981) - Posture is acceptable if not maintained or repeated
 * 3-4: Further Investigation (Amber #f59e0b) - Further investigation needed, changes may be required
 * 5-6: Change Soon (Orange #f97316) - Investigation and changes required soon
 * 7:   Immediate Action (Red #ef4444) - Investigation and changes required immediately
 */
export function evaluateRULA(
  cvaDeg: number,
  shoulderTiltDeg: number,
  trunkAngleDeg: number,
  ipdRatio: number = 1.0
): RulaEvaluation {
  let rulaScore: 1 | 2 | 3 | 4 | 5 | 6 | 7 = 1;

  // Base score from Craniovertebral Angle (CVA forward head posture)
  if (cvaDeg >= 50.0) {
    rulaScore = 1;
  } else if (cvaDeg >= 45.0) {
    rulaScore = 2;
  } else if (cvaDeg >= 40.0) {
    rulaScore = 4;
  } else if (cvaDeg >= 35.0) {
    rulaScore = 5;
  } else {
    rulaScore = 7;
  }

  // Adjust for excessive shoulder tilt asymmetry
  if (shoulderTiltDeg > 12.0) {
    rulaScore = Math.min(7, rulaScore + 2) as any;
  } else if (shoulderTiltDeg > 6.0) {
    rulaScore = Math.min(7, rulaScore + 1) as any;
  }

  // Adjust for trunk spine slouch/tilt
  if (trunkAngleDeg > 22.0) {
    rulaScore = Math.min(7, rulaScore + 2) as any;
  } else if (trunkAngleDeg > 14.0) {
    rulaScore = Math.min(7, rulaScore + 1) as any;
  }

  // Adjust for screen proximity slouch (leaning excessively close <45cm)
  if (ipdRatio > 1.35) {
    rulaScore = Math.min(7, rulaScore + 2) as any;
  } else if (ipdRatio > 1.25) {
    rulaScore = Math.min(7, rulaScore + 1) as any;
  }

  let level: RulaLevel;
  let status: PostureStatus;
  let colorHex: string;
  let actionRecommendation: string;

  if (rulaScore <= 2) {
    level = 'acceptable';
    status = 'Acceptable';
    colorHex = '#10b981'; // Emerald Green
    actionRecommendation = 'Optimal posture maintained. Keep spine upright and shoulders relaxed.';
  } else if (rulaScore <= 4) {
    level = 'investigate';
    status = 'Investigate';
    colorHex = '#f59e0b'; // Amber
    actionRecommendation = 'Slight slouch or forward head tilt detected. Gently tuck chin and align shoulders.';
  } else if (rulaScore <= 6) {
    level = 'change_soon';
    status = 'Change Soon';
    colorHex = '#f97316'; // Orange
    actionRecommendation = 'Pronounced posture fatigue. Sit upright, retract shoulders, and take a 20s break.';
  } else {
    level = 'immediate';
    status = 'Immediate Action';
    colorHex = '#ef4444'; // Rose Red
    actionRecommendation = 'Critical ergonomic strain. Pause and perform guided stretch session now.';
  }

  return {
    rula_score: rulaScore,
    level,
    status,
    color_hex: colorHex,
    action_recommendation: actionRecommendation,
  };
}
